Declarative Endpoint Traits + Specs (#4)
alextwoods/smithy: docs/source-2.0/aws/aws-endpoints.rst

.. _aws-endpoints:

===============================
AWS Declarative Endpoint Traits
===============================

> This document defines AWS Declarative endpoint traits.

Reviewer: ```suggestion
This document defines AWS declarative endpoint traits.
```

Reviewer: This section could also use a preamble discussing and linking to the concepts involved in AWS regions and endpoints.

.. smithy-trait:: aws.endpoints#endpointsModifier
.. _aws.endpoints#endpointsModifier-trait:

-----------------------------------------

> ``aws.endpoints#endpointsModifier`` trait

Reviewer: Can the summaries for traits that apply this trait be updated to lead with "An endpoints modifier trait that...", where `endpoints modifier trait` links to this section?

Author: done!

-----------------------------------------

Summary

>     A meta-trait that marks a trait as an endpoint modifier and describes the behavior

Reviewer: Cut here after the "and", as the next sentence includes the description of what traits with this meta-trait do.

    of endpoint resolution for services or operations.  Traits that are marked with this trait are
    applied to service shapes or operation shapes to indicate how a client can resolve
    endpoints for that service or operation.
Trait selector
    ``[trait|trait]``
Value type
    Annotation trait

The following example defines a service with ``standardRegionalEndpoints`` modified by
the hypothetical ``fooExample`` endpoint modifier.

.. code-block:: smithy

    $version: "2"

    namespace smithy.example

    use aws.endpoints#endpointsModifier
    use aws.endpoints#standardRegionalEndpoints

    @endpointsModifier
    @trait(selector: "service")
    structure fooExample {}

    @standardRegionalEndpoints
    @fooExample
    service MyService {
        version: "2020-04-02"
    }

> Because endpointsModifier definitions are just specialized shapes, they

Reviewer: ```suggestion
Because endpoint modification definitions are just specialized shapes, they
```

can also support configuration settings.

.. code-block:: smithy

    $version: "2"
    namespace smithy.example

    use aws.endpoints#endpointsModifier
    use aws.endpoints#standardRegionalEndpoints

    @endpointsModifier
    @trait(selector: "service")
    structure endpointSuffix {
        suffix: String
    }

    @standardRegionalEndpoints
    @endpointSuffix(suffix="-suffix")
    service MyService {
        version: "2020-04-02"
    }

.. smithy-trait:: aws.endpoints#standardRegionalEndpoints
.. _aws.endpoints#standardRegionalEndpoints-trait:

-------------------------------------------------
``aws.endpoints#standardRegionalEndpoints`` trait
-------------------------------------------------

Summary

>     Indicates that a services endpoints should be resolved using the standard regional

Reviewer: ```suggestion
    Indicates that a service's endpoints should be resolved using the standard AWS regional
```

    patterns:

>     - Default: ``{service}.{region}.{dnsSuffix}``

Reviewer: formatting: I think it does, but make sure this renders properly

Author: Yeah - looks like this renders correctly with the doc's`make html`

    - Fips: ``{service}-fips.{region}.{dnsSuffix}``
    - Dualstack: ``{service}.{region}.{dualStackDnsSuffix}``
    - Fips/Dualstack: ``{service}-fips.{region}.{dualStackDnsSuffix}``

Trait selector
    ``service``
Trait value
    A ``structure`` with the following properties:

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property
      - Type
      - Description
    * - partitionSpecialCases

>       - ``List<PartitionSpecialCase>`` of `PartitionSpecialCase object`_

Reviewer: formatting:

```suggestion
      - ``list`` of `PartitionSpecialCase object`_
```

      - A list of partition special cases - endpoints for a partition that do not follow the
        standard patterns.
    * - regionSpecialCases

>       - ``List<RegionSpecialCase>`` of `RegionSpecialCase object`_

Reviewer: formatting:

```suggestion
      - ``list`` of `RegionSpecialCase object`_
```

      - A list of regional special cases - endpoints for a region that do not follow the
        standard patterns.

Conflicts with
    :ref:`aws.endpoints#nonRegionalizedEndpoints-trait`

Most AWS services are regionalized and are strongly encouraged to follow
the standard endpoint patterns defined above, both for consistency and to

> ensure that endpoints are forwards compatible and that SDK updates are

Reviewer: ```suggestion
the standard endpoint patterns defined above for consistency, to
ensure that endpoints are forwards compatible, and that SDK updates are
```

not required when the service launches in a new region or partition.

The following example defines a service that uses the standard regional endpoints:

.. code-block:: smithy

    $version: "2"

    namespace smithy.example

    use aws.endpoints#standardRegionalEndpoints

    @standardRegionalEndpoints
    service MyService {
        version: "2020-04-02"
    }

While services are strongly encouraged to follow standard endpoint patterns,
there are occasional exceptions and special cases.  The following example defines
a service that use standard regional endpoints, but uses a non-standard pattern for
FIPS endpoints in US GovCloud:

.. code-block:: smithy

    @standardRegionalEndpoints{
        partitionSpecialCases: [
            {
                partition: "aws-us-gov",
                endpoint: "myservice.{region}.{dnsSuffix}",
                fips: true
            }
        ]
    }
    service MyService {
        version: "2020-04-02"
    }

> ---------------------------

Reviewer: Remove this line, will make this a sub-heading of the trait.

> PartitionSpecialCase object

Reviewer: ```suggestion
``PartitionSpecialCase`` object
```

---------------------------

> A PartitionSpecialCase object contains the following properties:

Reviewer: ```suggestion
A ``PartitionSpecialCase`` object contains the following properties:
```

Should also include a blurb on what this means for all regions in the partition.

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property name
      - Type
      - Description
    * - partition
      - ``string``

>       - **Required**. The partition to special case (example: "aws").

Reviewer: Would it be beneficial to have this as a map key instead of as a member of an object in a list?

Author: I had debated this a bit.  The issue is with variants (fips/dualstack) - you would then need to have the key map to a list of special cases, so the example above would become:
```
@standardRegionalEndpoints{
    partitionSpecialCases: {
      "aws-us-gov": [
          {
             endpoint: "myservice.{region}.{dnsSuffix}",
             fips: true
          }
       ]
   }
}
```

The same applies with the regionalSpecialCases.  We can have either 1) A list of special cases that include the region/partition or 2) a map with key of region/partition and values that are a list of special cases.  

I landed on 1 as most of our special cases are for a single variant (as in the example above).  But I don't have a strong preference between these two.  Thoughts?

    * - endpoint
      - ``string``
      - **Required**. The special cased endpoint template.
    * - dualStack
      - ``boolean``
      - When ``true`` the special case will apply to dualstack endpoint variants.
    * - fips
      - ``boolean``
      - When ``true`` the special case will apply to fips endpoint variants.

> ------------------------

Reviewer: Remove this line, will make this a sub-heading of the trait.

> RegionSpecialCase object

Reviewer: ```suggestion
``RegionSpecialCase`` object
```

------------------------

A RegionSpecialCase object contains the following properties:

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property name
      - Type
      - Description
    * - region
      - ``string``

>       - **Required**. The region to special case (example: "us-west-2").

Reviewer: Would it be beneficial to have this as a map key instead of as a member of an object in a list?

Author: See comment on PartitionSpecialCase - either way these two should be consistent I think.

    * - endpoint
      - ``string``
      - **Required**. The special cased endpoint template.
    * - dualStack
      - ``boolean``
      - When ``true`` the special case will apply to dualstack endpoint variants.
    * - fips
      - ``boolean``
      - When ``true`` the special case will apply to fips endpoint variants.
    * - signingRegion
      - ``string``
      - Override the signingRegion used for this region.

.. smithy-trait:: aws.endpoints#nonRegionalizedEndpoints
.. _aws.endpoints#nonRegionalizedEndpoints-trait:

-------------------------------------------------
``aws.endpoints#nonRegionalizedEndpoints`` trait
-------------------------------------------------

Summary
    Indicates that a service is non-regionalized and a single endpoint should be resolved
    per partition.

> 

Reviewer: Remove empty line

Trait selector
    ``service``
Trait value
    A ``structure`` with the following properties:

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property
      - Type
      - Description
    * - endpointPattern
      - ``string``

>       - **Required** The pattern to use for the partition endpoint.  This value can be set to ``service_dnsSuffix`` to

Reviewer: Are these values special cased and this can be anything, but these are shortcuts? Or are they they only valid values?

Author: This is intended to be an enum and those are the only two valid values.  Is there a prefered way to document it?  I was referencing `httpApiKeyAuth` trait's "in" member, which just has "This value can be set to header or query."

Reviewer: I think the wording just threw me off - It's a type not the explicit pattern. "The pattern type to use for the partition endpoint", since it's not the literal pattern to use.

        use the ``{service}.{dnsSuffix}`` pattern or ``service_region_dnsSuffix`` to use
        ``{service}.{region}.{dnsSuffix}``.
    * - partitionEndpointSpecialCases

>       - ``List<PartitionEndpointSpecialCase>`` of `PartitionEndpointSpecialCase object`_

Reviewer: formatting:

```suggestion
      - ``list`` of `PartitionEndpointSpecialCase object`_
```

      - A list of partition endpoint special cases - partitions that do not follow the
        services standard patterns or are located in a region other than the partition's
        ``defaultGlobalRegion``.

Conflicts with
    :ref:`aws.endpoints#standardRegionalEndpoints-trait`

> Non-regionalized (also known as "global" services) resolve a single endpoint per partition.

Reviewer: Link from `non-regionalized` here if there's good, public documentation.

- ``service_dnsSuffix``: ``{service}.{dnsSuffix}``
- ``service_region_dnsSuffix``: ``{service}.{region}.{dnsSuffix}``

The following example defines a non-regionalized service that uses ``{service}.{dnsSuffix}``:

.. code-block:: smithy

    $version: "2"

    namespace smithy.example

    use aws.endpoints#nonRegionalizedEndpoints

    @nonRegionalizedEndpoints(endpointPattern: "service_dnsSuffix")
    service MyService {
        version: "2020-04-02"
    }

Services should follow the standard patterns; however, occasionally there are special cases.
The following example defines a non-regionalized service that uses a special case pattern in
the ``aws`` partition and uses a non-standard global region in the ``aws-cn`` partition:

.. code-block:: smithy

    @nonRegionalizedEndpoints {
        endpointPattern: "service_dnsSuffix",
        partitionEndpointSpecialCases: [
            {
                partition: "aws",
                endpoint: "myservice.global.amazonaws.com"
            },
            {
                partition: "aws-cn",
                region: "cn-north-1"
            }
        ]
    }
    service MyService {
        version: "2020-04-02"
    }

> -----------------------------------

Reviewer: Remove this line, will make this a sub-heading of the trait.

> PartitionEndpointSpecialCase object

Reviewer: ```suggestion
``PartitionEndpointSpecialCase`` object
```

-----------------------------------

A PartitionEndpointSpecialCase object contains the following properties:

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property name
      - Type
      - Description
    * - partition
      - ``string``
      - **Required**. The partition to special case (example: "aws").
    * - endpoint
      - ``string``
      - The special cased endpoint template.
    * - region
      - ``string``
      - Override the ``defaultGlobalRegion`` used in this partition.

.. smithy-trait:: aws.endpoints#dualStackOnlyEndpoints
.. _aws.endpoints#dualStackOnlyEndpoints-trait:

----------------------------------------------
``aws.endpoints#dualStackOnlyEndpoints`` trait
----------------------------------------------

Summary

>     Indicates that a service has only dualStack endpoints and should not

Reviewer: Link from `dualStack` here if there's good, public documentation.

    have the ``useDualStackEndpoint`` endpoint parameter.

> 

Reviewer: Remove empty line

Trait selector
    ``service``
Trait value
    Annotation trait

Adding the dualStackOnlyEndpoints to a service modifies the generation of endpoints from

> _aws.endpoints#standardRegionalEndpoints-trait or _aws.endpoints#nonRegionalizedEndpoints-trait

Reviewer: formatting: make sure these render properly.

Author: Thanks - fixed.

and removes the ``useDualStackEndpoint`` parameter and defaults the behavior to dualstack for
all partitions that support it.

The following example specifies a service that uses standard regional endpoint patterns and
is dualStack only:

.. code-block:: smithy

     @standardRegionalEndpoints
     @dualStackOnlyEndpoints
     service MyService {
         version: "2020-04-02"
     }

.. smithy-trait:: aws.endpoints#rulesBasedEndpoints
.. _aws.endpoints#rulesBasedEndpoints-trait:

-------------------------------------------
``aws.endpoints#rulesBasedEndpoints`` trait
-------------------------------------------

Summary
    Indicates that a service has hand written endpoint rules.

> 

Reviewer: Remove empty line.

Trait selector
    ``service``
Trait value
    Annotation trait

Services marked with the ``rulesBasedEndpoints`` trait have hand written endpoint rules that
extend or replace their standard generated endpoint rules.  This trait marks the presence
of hand written rules, which may be added to the model by a transformer,
but does not specify their behavior.

> The following example specifies a service that has standard regional endpoints extended with

Reviewer: Is there a need to define replace vs extend? How does extending work functionally?

Author: Added a bit more detail.

Reviewer: I'm still not quite sure what this is saying and it would be clear to me based on this if the implementation of this trait, other traits, or something else entirely is responsible for the functionality described here. Can there be conflicts and how are they resolved? Is there an order of precedence?

A PartitionEndpointSpecialCase object contains the following properties:

.. list-table::
    :header-rows: 1
    :widths: 10 30 60

    * - Property name
      - Type
      - Description
    * - partition
      - ``string``
      - **Required**. The partition to special case (example: "aws").
    * - endpoint
      - ``string``
      - The special cased endpoint template.
    * - region
      - ``string``
      - Override the ``defaultGlobalRegion`` used in this partition.

.. smithy-trait:: aws.endpoints#dualStackOnlyEndpoints
.. _aws.endpoints#dualStackOnlyEndpoints-trait:

----------------------------------------------
``aws.endpoints#dualStackOnlyEndpoints`` trait
----------------------------------------------

Summary
    An :ref:`endpoints modifier trait <aws.endpoints#endpointsModifier-trait>`
    that indicates that a service has only
    `dual stack endpoints <https://docs.aws.amazon.com/general/latest/gr/rande.html#dual-stack-endpoints>`_,
    does not support IPV4 only endpoints, and should not have the ``useDualStackEndpoint`` endpoint parameter.
    Dual stack endpoints support IPV4 and IPV6.
Trait selector
    ``service``
Trait value
    Annotation trait

> Adding the dualStackOnlyEndpoints to a service modifies the generation of endpoints from

Reviewer: ```suggestion
Adding the ``dualStackOnlyEndpoints`` trait to a service modifies the generation of endpoints from
```

        standard patterns.
    * - regionSpecialCases
      - ``map`` of region to `RegionSpecialCase object`_
      - A map of region to regional special cases - endpoints for a region that do not follow the
        standard patterns.

Conflicts with
    :ref:`aws.endpoints#standardPartitionalEndpoints-trait`

Most AWS services are regionalized and are strongly encouraged to follow
the standard endpoint patterns defined above for consistency, and to
ensure that endpoints are forwards compatible, and that SDK updates are
not required when the service launches in a new region or partition.

The following example defines a service that uses the standard regional endpoints:

.. code-block:: smithy

    $version: "2"

    namespace smithy.example

    use aws.endpoints#standardRegionalEndpoints

    @standardRegionalEndpoints
    service MyService {
        version: "2020-04-02"
    }

While services are strongly encouraged to follow standard endpoint patterns,
there are occasional exceptions and special cases.  The following example defines
a service that use standard regional endpoints, but uses a non-standard pattern for
FIPS endpoints in US GovCloud:

.. code-block:: smithy

>     @standardRegionalEndpoints{

Reviewer 2: ```suggestion
    @standardRegionalEndpoints(
```

>         partitionSpecialCases: {

Reviewer 2: ```suggestion
        partitionSpecialCases: (
```

            aws-us-gov: [
                {
                    endpoint: "myservice.{region}.{dnsSuffix}",
                    fips: true
                }
            ]
        }

>     }

Reviewer 2: ```suggestion
    )
```
